package/kedro_viz/models/flowchart.py

```diff
@@ -416,52 +429,54 @@ def __post_init__(self, task_node: TaskNode):
             )
 
 
-# pylint: disable=too-many-instance-attributes
-@dataclass
 class DataNode(GraphNode):
     """Represent a graph node of type DATA"""
 
     # whether the data node is a free input
-    is_free_input: bool = field(default=False)
+    is_free_input: bool = False
 
     # the layer that this data node belongs to
-    layer: Optional[str] = field(default=None)
+    layer: Optional[str]
 
     # the concrete type of the underlying kedro_obj
-    dataset_type: Optional[str] = field(init=False)
+    dataset_type: Optional[str]
 
     # the list of modular pipelines this data node belongs to
-    modular_pipelines: List[str] = field(init=False)
+    modular_pipelines: List[str] = []
 
-    viz_metadata: Optional[Dict] = field(init=False)
+    viz_metadata: Optional[Dict]
 
     # command to run the pipeline to this node
-    run_command: Optional[str] = field(init=False, default=None)
+    run_command: Optional[str]
 
     # the type of this graph node, which is DATA
     type: str = GraphNodeType.DATA.value
 
     # statistics for the data node
-    stats: Optional[Dict] = field(default=None)
+    stats: Optional[Dict]
 
-    def __post_init__(self):
+    def __init__(self, **data):
+        super().__init__(**data)
         self.dataset_type = get_dataset_type(self.kedro_obj)
 
         # the modular pipelines that a data node belongs to
         # are derived from its namespace, which in turn
         # is derived from the dataset's name.
         self.namespace = self._get_namespace(self.name)
         self.modular_pipelines = self._expand_namespaces(self._get_namespace(self.name))
-        metadata = getattr(self.kedro_obj, "metadata", None)
-        if metadata:
-            try:
+
+        try:
+            metadata = getattr(self.kedro_obj, "metadata", None)
+
+            if isinstance(metadata, dict):
                 self.viz_metadata = metadata["kedro-viz"]
-            except (AttributeError, KeyError):  # pragma: no cover
-                logger.debug("Kedro-viz metadata not found for %s", self.name)
+
+        except (AttributeError, KeyError):  # pragma: no cover
+            logger.debug("Kedro-viz metadata not found for %s", self.name)
```

Comment:
I don't find this refactoring is better than the previous form.  Original Line 455 is using `getattr` already and shouldn't be buried in a try-except block. 

In fact I found the original logic weird too (What could lead to `AttributeErorr`? and if metadata is not a dict you get `TypeError` which isn't caught). I will either use try-except to catch the whole metadata block or use `getattr`

Since this PR focus on dataclass refactoring I suggest leave it and don't make this change.

Reply:
There was an issue with linting. If we try to access metadata["kedro-viz"] without the isinstance check. So I added that. I removed the getattr from try block now, since it has a default None value